File: src/main/java/com/freya02/botcommands/api/localization/LocalizationTemplate.java
package com.freya02.botcommands.api.localization;

import org.jetbrains.annotations.NotNull;

/**
 * Allows different implementation of localization templates.
 * <br>Those templates can have different specifications if you wish, but the default one is in {@link DefaultLocalizationTemplate}.
 *
 * @see DefaultLocalizationTemplate
 */
public interface LocalizationTemplate {
	/**
	 * Processes the localization template and replaces the named parameters by theirs values
	 */
	@NotNull String localize(Localization.Entry... args);
}

File: src/main/kotlin/com/freya02/botcommands/api/localization/LocalizationTemplate.kt
package com.freya02.botcommands.api.localization

/**
 * Allows different implementation of localization templates.
 *
 * A localization template represents the entire string being localized, with parameters.
 *
 * You can implement your own template, the default one is [DefaultLocalizationTemplate].
 *
 * @see DefaultLocalizationTemplate
 */
interface LocalizationTemplate {
    /**
     * Processes the localization template and replaces the named parameters by theirs values
     */
    fun localize(vararg args: Localization.Entry): String
}

/**
 * Processes the localization template and replaces the named parameters by theirs values
 */
fun LocalizationTemplate.localize(vararg args: Pair<String, Any>): String =
    localize(*args.map { (k, v) -> Localization.Entry.entry(k, v) }.toTypedArray())